torch/distributed/_functional_collectives.py

```diff
@@ -0,0 +1,196 @@
+from typing import Any, Tuple, Union, List, cast
+
+import torch
+from torch.autograd import Function
+from torch.utils.weak import WeakIdKeyDictionary
+import torch.distributed as dist
+
+from torch._C import _disabled_torch_function_impl
+from torch.utils._pytree import tree_map
+
+import torch.distributed.distributed_c10d as c10d
+import torch.distributed._tensor as dt
+
+"""
+New traceable, functional collectives.
+  compiler: trace these ops with plain-old-data schemas, then choose how to lower them.
+  eager: execute these 'functional' ops which in eager return AsyncCollectiveTensor subclasses,
+         automatically calling .wait() on underlying/hidden async 'work' obj only when fed to
+         a downstream op.
+
+Issues:
+* Where should these ops live? Couldn't `import torch` if putting these ops in existing torch.distributed files
```

Comment:
I think a tricky issue is that some build without distributed still could access the aten op and probably this file too, if they use it, it wouldn't work because pg is not built.